### PR TITLE
Tempo: Carry over query from Search tab to TraceQL tab

### LINKED ISSUE #79038: Carry over TraceQL query when going from "Search" to "TraceQL" tab in Tempo Datasource Explore
**Why is this needed**:

As a user, I may start by creating a TraceQL query in the "Search" tab of the Tempo datasource explore. This is a nice graphical way to create queries. However, sometimes I may want to switch over to the raw TraceQL editor view from there to modify my query further. 

Today, I lose whatever query I'd built in the "Search" view when I make the switch, which is frustrating because it means I have to start over from scratch in the "TraceQL" tab. This is especially frustrating because there is no warning I'm going to lose my work in the "Search" tab. 

Today the only workaround is to manually highlight my query in the "Search" view and then copy paste it into the "TraceQL" view. 

**What would you like to be added**:

I'd like Grafana itself to carry over the query I'd created when I switch over to the "TraceQL" tab so I don't lose my work. 

**Who is this feature for?**

Users who are moving between the TraceQL and Search tabs in the Tempo datasource explore. Those users may want to start with the visual query builder, but then want to move to the TraceQL view since it allows more flexibility in defining and crafting their query.

### PR DESCRIPTION
**Pull request by joey-grafana**
**What is this feature?**

Carries over query from Search tab to TraceQL tab.

**Why do we need this feature?**

So users can easily construct their query and then add more advanced operations if needed (i.e. avoid losing what they've already constructed in the `Search` tab).

**Who is this feature for?**

Tempo users.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/79038

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by joey-grafana** [comment]
There are two issues in question here;
1) Carrying over the query from the Search tab to the TraceQL tab (issue linked to this PR)
   - As discussed last week this was to be worked on.
2) Carrying over the query from the TraceQL tab to the Search tab ([issue here](https://github.com/grafana/grafana/issues/79039))
   - This was to be added to the backlog. This also consists of possibly two parts. One being lossy conversion of the TraceQL query when moving from the TraceQL tab to the Search tab. The second part being total conversion of the TraceQL query when moving from the TraceQL tab to the Search tab (which is actually quite difficult to get right).

> New proposal: when in the TraceQL tab show an alert/CTA to continue editing the query from the Search tab (if there are values in it, check the filters variable of TempoQuery). Two options to achieve it:

This issue is only for moving from the Search tab to the TraceQL tab (issue 1 above).

> I think this will not solve the pain point of users navigating from Search to TraceQL and "expecting" the query to carry over since the button is a bit out of the way and camouflaged with other actions.

Little bit surprised as this is essentially what we agreed on last week. If we want to improve the UX that is fair but if I am not sure the above does that (if you did mean that it is for moving from Search to TraceQL and not from TraceQL to Search).

- Reason being is that the banner should only show when there is actually values to copy over, right? That means it's hidden when there's no values. So as soon as you change anything in the Search tab, the entire content will shift down as the banner displays, IMO that's poorer UX than just having a button to copy over the query. 
- Also, I think we should be conservative with our vertical space in the Search tab.
- Many users who are not interested in writing TraceQL but prefer the builder will also not want to copy over their query so the banner will always be in the way.

I like the idea of a banner but I think this is more suited to moving from the TraceQL tab to the search tab, which will be in a different issue (issue 2 above) and can describe that the process is lossy until we have a better approach where we can copy over the entire query to the Search tab.

**Comment by joey-grafana** [comment]
This is good however users won't know they can copy over the query until they go to the TraceQL tab. Users would have to dismiss this banner every time they go to the TraceQL editor if any part of the Search tab is selected. I think the benefit of the banner is that we are being more explicit about what the action is doing but the button itself will not add extra visual complexity to the TraceQL tab. 

I'm going to bring this to the UX feedback session!

**Comment by joey-grafana** [comment]
Ok so brought this to the UX feedback session :)

There was one suggestion to change the banner in the TraceQL tab to remove the query and only contain the text `Continue editing the query from Search?` with a button next to it. So it would look something like;
Screenshot 2023-12-13 at 14 57 15

The button should have a better text though as this seems repetitive. 

We would also have to check that the query in search is actually different than the one in TraceQL and if it is then show this banner else don't render it.

I will say that more people were in favor of option one, where we show the Copy to TraceQL button, with one caveat, it would be better if it was closer to the Search editor so it's purpose is clearer. One way to improve this would be to add a tooltip to the button as discussed in the UX feedback.

**Comment by joey-grafana** [comment]
> small banner above the query editor "Continue editing the query from Search? { span.attr1 = "foo" && resource.attr2 = 123}. Clicking it would copy the query to the editor and hide the banner.

Ahhh right I was assuming you wanted the query from the example above in the text but it was only an example. I've updated the user flow anyway so it's on the TraceQL tab and so the banner disappears after copying the query. I think this works well for the user.

**Review by adrapereira — CHANGES_REQUESTED** [review]
Thank you for the changes! Left a couple of small comments.

File: public/app/plugins/datasource/tempo/traceql/QueryEditor.tsx [hunk 1]
```diff
@@ -29,6 +44,34 @@ export function QueryEditor(props: Props) {
           Documentation
         </a>
       </InlineLabel>
+      {showCopyFromSearchButton && (
+        <InlineLabel>
+          <div>
+            Continue editing the query from the Search tab?
```
[the remark this answers is not in the document]
Comment (joey-grafana) [reply public/app/plugins/datasource/tempo/traceql/QueryEditor.tsx:45]: @knylander-grafana could you please let us know what you think of this copy and if it should be updated when you get back from holidays?

@adrapereira I agree and think this looks good for now (but we will also check the text for improvements) so would like to get this merged before the weekend.

**Review by adrapereira — APPROVED** [review]

**Pull request merged**